src/Build/Instance/TaskRegistry.cs

```diff
@@ -1126,6 +1099,7 @@ internal RegisteredTaskRecord(string registeredName, AssemblyLoadInfo assemblyLo
                 _taskFactoryParameters = taskFactoryParameters;
                 _taskIdentity = new RegisteredTaskIdentity(registeredName, taskFactoryParameters);
                 _parameterGroupAndTaskBody = inlineTask;
+                _registrationOrderId = Interlocked.Increment(ref s_nextRegistrationOrderId);
```

Comment:
In light of multi-threading, is it guaranteed that the order defined by this global monotonically increasing counter is always stable in all scenarios? I.e. can multiple threads be registering tasks from, say, different project files, each defining a task with a colliding name, and racing to define the one in their file as "first"?

Reply:
I'm specifically thinking of cases where there is no race in the user input but MSBuild internally decides to do some work in parallel. It looks unlikely but not all that far-fetched either.

Reply:
Great question!
Can we actually have `UsingTask` that step outside of the project boundary?

*I believe* we are entitled to keep undeterministic behavior in such (theoretical only?) scenario (multple project files with coliding UsingTask elements) - as otherwise we'd need to commit to zero level of paralelism or add ability to explicitly specify priority/order of tasks. If explicit order would need to be added by user - they can already add disambiguate namings.

Again - I beleive we are concerned only in a single project here - so the situation likely doesn't exist